File: TESS_CppAPI_EXAMPLE/MyPlugin.cpp
#include "MyPlugin.h"

#include <QDockWidget>
#include <QMessageBox>

#include "tessinterface.h"
#include "guiinterface.h"
#include "MyNet.h"
#include "MySimulator.h"
#include "TESS_API_EXAMPLE.h"
#include "Plugin/StepsPerCall.h"

MyPlugin::MyPlugin()
	:mpExampleWindow(nullptr), mpMyNet(nullptr), mpMySimulator(nullptr)
{
}

void MyPlugin::init()
{
	initGui();

	mpMyNet = new MyNet();
	mpMySimulator = new MySimulator();

	QObject::connect(mpMySimulator, SIGNAL(forRunInfo(const QString)), mpExampleWindow, SLOT(showRunInfo(const QString)));

	QMainWindow *pMainWindow = gpTessInterface->guiInterface()->mainWindow();
	QObject::connect(mpMySimulator, SIGNAL(forReStartSimu()), pMainWindow, SLOT(doStartSimu()), Qt::QueuedConnection);
	QObject::connect(mpMySimulator, SIGNAL(forStopSimu()), pMainWindow, SLOT(doStopSimu()), Qt::QueuedConnection);
}

void MyPlugin::initGui() {
	//在TESS NG主界面上增加 QDockWidget对象
	mpExampleWindow = new TESS_API_EXAMPLE();
	QDockWidget *pDockWidget = new QDockWidget(QString("自定义与TESS NG交互界面"), (QWidget *)gpTessInterface->guiInterface()->mainWindow());
	pDockWidget->setObjectName(QStringLiteral("mainDockWidget"));
	pDockWidget->setFeatures(QDockWidget::NoDockWidgetFeatures);
	pDockWidget->setAllowedAreas(Qt::LeftDockWidgetArea);
	pDockWidget->setWidget(mpExampleWindow->centralWidget());
	gpTessInterface->guiInterface()->addDockWidgetToMainWindow(static_cast<Qt::DockWidgetArea>(1), pDockWidget);

	//增加菜单及菜单项
	QMenuBar *pMenuBar = gpTessInterface->guiInterface()->menuBar();
	QMenu *pMenu = new QMenu(pMenuBar);
	pMenu->setObjectName(QString::fromUtf8("menuExample"));
	pMenuBar->addAction(pMenu->menuAction());
	pMenu->setTitle(QString("范例菜单"));

	QAction *pActionOk = new QAction(pMenuBar->parent());
	pActionOk->setObjectName("actionExample");
	pActionOk->setText(QString("范例菜单项"));
	pActionOk->setCheckable(true);

	pMenu->addAction(pActionOk);

	QObject::connect(pActionOk, &QAction::triggered, []() {
		QMessageBox::information(gpTessInterface->guiInterface()->mainWindow(), QString("提示"), QString("is ok!"));
	});

}

void MyPlugin::unload() {
	delete mpMyNet;
	delete mpMySimulator;
	delete mpExampleWindow;
}

CustomerGui *MyPlugin::customerGui()
{
	return nullptr;
}

CustomerNet *MyPlugin::customerNet()
{
	return mpMyNet;
}

CustomerSimulator *MyPlugin::customerSimulator()
{
	return mpMySimulator;
}

MyPlugin::~MyPlugin()
{
}

File: TESS_PythonAPI_demo/17网联车辆汇入决策/MyPlugin.py
# -*- coding: utf-8 -*-

from PySide2.QtWidgets import QDockWidget
from Tessng import TessPlugin
from MyNet import *
#from MySimulator_cv_platoon import *
from MySimulator_cv_merge import *
from TESS_API_EXAMPLE import *

# 用户插件，继承自TessPlugin
class MyPlugin(TessPlugin):
    def __init__(self):
        super(MyPlugin, self).__init__()
        self.mNetInf = None
        self.mSimuInf = None

    def initGui(self):
        # 在TESS NG主界面上增加 QDockWidget对象
        self.examleWindow = TESS_API_EXAMPLE()

        iface = tngIFace()
        win = iface.guiInterface().mainWindow()

        dockWidget = QDockWidget("自定义与TESS NG交互界面", win)
        dockWidget.setObjectName("mainDockWidget")
        dockWidget.setFeatures(QDockWidget.NoDockWidgetFeatures)
        dockWidget.setAllowedAreas(Qt.LeftDockWidgetArea)
        dockWidget.setWidget(self.examleWindow.centralWidget())
        iface.guiInterface().addDockWidgetToMainWindow(Qt.DockWidgetArea(1), dockWidget)

    # 过载父类方法，在 TESS NG工厂类创建TESS NG对象时调用
    def init(self):
        self.initGui()
        self.mNetInf = MyNet()
        self.mSimuInf = MySimulator()
        self.mSimuInf.signalRunInfo.connect(self.examleWindow.showRunInfo)
        iface = tngIFace()
        win = iface.guiInterface().mainWindow()
        #将信号mSimuInf.forReStartSimu关联到主窗体的槽函数doStartSimu，可以借实现自动重复仿真
        self.mSimuInf.forReStartSimu.connect(win.doStartSimu, Qt.QueuedConnection)

    # 过载父类方法，返回插件路网子接口，此方法由TESS NG调用
    def customerNet(self):
        return self.mNetInf

    #过载父类方法，返回插件仿真子接口，此方法由TESS NG调用
    def customerSimulator(self):
        return self.mSimuInf
